### PR TITLE
[LLD] Tombstone LocalTU entry in .debug_names

### PR DESCRIPTION
**Pull request by ayermolo**
By default LLD uses 0x0... as a tombstone value. For Type Units this is a valid
address. Changed so that UINT{32,64}_MAX value is used for tombstone. The
former is for DWARF32, and latter for DWARF64. The default for LLVM is 32 bit DWARF.
Although 64 bit DWARF is supported in LLVM, it's not widely used. The value was chosen
because it is not a valid offset.

This was introduced as a change to DWARF6 spec:
https://dwarfstd.org/issues/231013.1.html

### CODE REVIEW

**Comment by ayermolo** [comment]
Side note. For DWARF64 the check on line 921 is not necessary. Do I need to change relocation in https://github.com/llvm/llvm-project/pull/70515?

File: lld/ELF/InputSection.cpp [hunk 1]
```diff
@@ -918,7 +917,8 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
       continue;
 
     if (tombstone ||
-        (isDebug && (type == target.symbolicRel || expr == R_DTPREL))) {
+        ((isDebug && (type == target.symbolicRel || expr == R_DTPREL))) ||
```
Comment (dwblaikie) [inline lld/ELF/InputSection.cpp:920]: Would be nice to know from @MaskRay (or anyone else) why this `type == target.symbolicRel` is here, and if there's some way we can generalize it to cover the `.debug_names` case we're interested in (a `R_X86_64_32` relocation) without needing to special case `isDebugNames` in this condition

File: lld/ELF/InputSection.cpp [hunk 2]
```diff
@@ -947,6 +947,22 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
...
+        switch (type) {
+        case R_X86_64_64:
+          maxVal = llvm::maxUIntN(64);
+          break;
+        case R_X86_64_32:
+          maxVal = llvm::maxUIntN(32);
+          break;
+        default:
+          llvm_unreachable("Unsupported relocation type in .debug_names.");
+        }
```
Comment (dwblaikie) [inline lld/ELF/InputSection.cpp:962]: Be nice to hear from @MaskRay or otherwise about whether we can generalize this in some way.

@ayermolo perhaps you can check the commit history to see how the `-1 as tombstone` worked - it was committed a few years back, so there might be some hint about how to do that that doesn't require special casing each relocation

File: lld/ELF/InputSection.cpp (hunk 1 above)
Comment (ayermolo) [reply lld/ELF/InputSection.cpp:920]: Yes. :D As I mentioned, for DWARF64 isDebugNames check is not necessary.

File: lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s [hunk 3]
```diff
@@ -0,0 +1,397 @@
+	.text
```
Comment (dwblaikie) [inline lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s:1]: I imagine if you check the other test cases (for things like the DebugLoc/DebugRanges cases) you'll find they don't use full DWARF but probably just some relocations in an otherwise empty/small section - and we should probably test this in a similar way. (if the DebugLoc/DebugRanges tests do use fully formed DWARF, that'll be a surprise to me, but then happy to work with you to see if we can simplify the test(s) as much as possible in that form)

Comment (ayermolo) [reply lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s:1]: Yeah I can look into simplifying.

File: lld/ELF/InputSection.cpp (hunk 1 above)
Comment (dwblaikie) [reply lld/ELF/InputSection.cpp:920]: While we're waiting for an expert opinion, if youre bored, you could try removing the `type == target.symbolicRel` check and running the lld tests, about 5 tests fail - investigating one fo those failures to see what goes wrong in the absence of that condition might help point the way to understanding how to generalize the condition enough to cover the `.debug_names` situation, without generalizing too far and breaking those tests

Comment (ayermolo) [reply lld/ELF/InputSection.cpp:920]: Well basically that condition checks if relocation is 64bit: R_X86_64_64 (type == 1).
Primarily used in .debug_addr. Incidentally why DWARF64 tests works without adding isDebugNames to that condition. since for .debug_names relocation type is controlled by DWARF32 vs DWARF64.
Slightly more interesting case is expr == R_DTPREL. Since pretty much all relocs in debug sections are ABS. Looks like that relocation is used to reference TLS symbols. Learned something new today.

Comment (dwblaikie) [reply lld/ELF/InputSection.cpp:920]: Yeah, mechanically I understand what the check does. 

What I mean is: Why is this check necessary? What bad things happen (some tests fail, but I haven't looked closely at what they're testing or what the bad behavior looks like) when this check fails/isn't checked? And why is it OK to ignore this check in the `.debug_names` case? And could we look at that reason and generalize the check so it's not specifically about `.debug_names`, but about whatever general property is true in the `.debug_names` case and would be true in other cases?

Basically, it seems the `type == target.symbolicRel` check is overly restrictive (we have at least one counter example - the `.debug_names` `R_X86_64_32` situation) and I'd like to know ideally how to generically relax this check to account for our situation, but also others that may come up in the future/just not to have a weird `.debug_names` special case that we don't really understand/can't explain why it's acceptable here, but not there - and layering special cases makes the code harder to understand and maintain.

Comment (ayermolo) [reply lld/ELF/InputSection.cpp:920]: Ah, I see what you mean. Yeah was going to dig in further today and see how to generalize. You are right it is over restrictive. Ideally it probably should be something like if (tombstone || debug), and then some logic depending on relocation type and probably section. I am hesitant to change current defaults due to downstream tools (cough lldb), breaking or something.

**Comment by ayermolo** [comment]
Updated with more generic version. Step in right direction?

File: lld/ELF/InputSection.cpp [hunk 4]
```diff
@@ -896,6 +893,16 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
...
+  const uint64_t debugTombstone = StringSwitch<uint64_t>(name)
+                                      .Case(".debug_ranges", 1)
+                                      .Case(".debug_loc", 1)
+                                      .Case(".debug_names", llvm::maxUIntN(32))
+                                      .Default(0);
+  // If -z dead-reloc-in-nonalloc= is specified, respect it.
+  if (!tombstone && isDebug)
+    tombstone = debugTombstone;
+  else if (tombstone)
+    tombstone = SignExtend64<bits>(*tombstone);
```
Comment (dwblaikie) [inline lld/ELF/InputSection.cpp:905]: if this is sign extended to 64 bits, but then we have a 32 bit relocation, I'd expect this fails/errors out?

File: lld/ELF/InputSection.cpp [hunk 5]
```diff
@@ -947,12 +953,14 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
       //
       // TODO To reduce disruption, we use 0 instead of -1 as the tombstone
       // value. Enable -1 in a future release.
+
+      // Extend to 64bit MAX for 64 bit relocations, LocalTU, in .debug_names.
+      if (isDebugNames && type == target.symbolicRel)
+        tombstone = SignExtend64<32>(*tombstone);
```
Comment (dwblaikie) [inline lld/ELF/InputSection.cpp:959]: I guess this only handles one case of 64 bit relocations, but there might be others? & I'm not sure `isDebugNames` is the special case here - shouldn't it be known by the relocation, rather than the section the relocation is in?

**Comment by dwblaikie** [comment]
> Updated with more generic version. Step in right direction?

I think it's probably going in the right direction - though I'm not sure we'll be quite able to figure out the right thing without some more lld expertise. But trying to make what progress we can in the interim.

File: lld/ELF/InputSection.cpp (hunk 4 above)
Comment (ayermolo) [reply lld/ELF/InputSection.cpp:905]: It is the same as original logic. It would go into if condition if tombstone is specified, and always extend.
bits == 64
So at least on ELF64 this is no-op. So TBH I am not sure why it's there.

```
template <unsigned B> constexpr inline int64_t SignExtend64(uint64_t x) {
  static_assert(B > 0, "Bit width can't be 0.");
  static_assert(B <= 64, "Bit width out of range.");
  return int64_t(x << (64 - B)) >> (64 - B);
}
```

File: lld/ELF/InputSection.cpp (hunk 5 above)
Comment (ayermolo) [reply lld/ELF/InputSection.cpp]: ok, made it more generic.

**Comment by ayermolo** [comment]
> > Updated with more generic version. Step in right direction?
> 
> I think it's probably going in the right direction - though I'm not sure we'll be quite able to figure out the right thing without some more lld expertise. But trying to make what progress we can in the interim.

Anyone else we can rope into this?

File: lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s (hunk 3 above)
Comment (ayermolo) [reply lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s]: Simplified, or does it need to be even further that it's not valid .debug_names section, and just has reloc?

**Comment by MaskRay** [comment]
Sorry for my late reply. https://github.com/llvm/llvm-project/pull/68131 IIUC `clang -fdebug-types-section -g -gpubnames` does not emit `.debug_names`. #68131 will implement the support (accidentally committed, reverted in f320065aeb379d8f9d2780a687768742f8b18eb2). `git cherry-pick -n 9bbd2bf654634cd95dd0be7948ec8402c3c76e1e` has merge conflicts, so it seems difficult for me to play with the example...

**Comment by MaskRay** [comment]
> By default LLD uses 0x0... as a tombstone value. For Type Units this is a valid address. Changed so that MAX value is used for tombstone.
> 
> This was introduced as a change to DWARF6 spec: [dwarfstd.org/issues/231013.1.html](https://dwarfstd.org/issues/231013.1.html)

The description (the draft commit message when the patch eventually lands) can use some clang commands for the curious to play with.

Some information can be extracted from 
@dwblaikie's https://discourse.llvm.org/t/debug-names-tombstoning-for-local-type-units/74410/2

```
$ clang++-tot null.cpp a.cpp b.cpp -g -fdebug-types-section -fuse-ld=lld -gpubnames && llvm-dwarfdump-tot a.out -debug-names
    LocalTU[0]: 0x00000030
 # some comments to explain what this is and how it is undesired can be useful
    LocalTU[0]: 0x00000000
```

File: lld/ELF/InputSection.cpp [hunk 6]
```diff
@@ -917,8 +925,7 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
     if (expr == R_NONE)
       continue;
 
-    if (tombstone ||
-        (isDebug && (type == target.symbolicRel || expr == R_DTPREL))) {
+    if (tombstoneValueToUse) {
```
Comment (MaskRay) [inline lld/ELF/InputSection.cpp:928]: I think
`if (!sym.getOutputSection() || (ds && ds->folded && !isDebugLine))` should be the outer `if` and the tombstone logic should be the inner `if`.

**Comment by ayermolo** [comment]
> Sorry for my late reply. IIUC `clang -fdebug-types-section -g -gpubnames` does not emit `.debug_names`. #68131 will implement the support (accidentally committed, reverted in [f320065](https://github.com/llvm/llvm-project/commit/f320065aeb379d8f9d2780a687768742f8b18eb2)). `git cherry-pick -n 9bbd2bf654634cd95dd0be7948ec8402c3c76e1e` has merge conflicts, so it seems difficult for me to play with the example...
Sorry I thought I mentioned somewhere.
https://github.com/llvm/llvm-project/pull/70515 Is a current open PR for it. At least according to GH there are no merge conflicts for it.

File: lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s (hunk 3 above)
Comment (dwblaikie) [reply lld/test/ELF/Inputs/dwarf5-64-debug-names-type-comdat-helper.s]: I'd go further, not making it a valid .debug_names section - like the tests in debug-dead-reloc.s (or it could even be added to that - it tests comdat .text already, so it could test comdat .debug_types too) - that tests special cases for .debug_loc and .debug_ranges, for instance, but doesn't have whole valid loclists or range lists.

(but mostly whatever @maskray's happy with)

**Comment by ayermolo** [comment]
@MaskRay  ping :)

File: lld/test/ELF/Inputs/dwarf5-debug-names-type-comdat.s [hunk 7]
```diff
@@ -0,0 +1,26 @@
+	.section	.debug_info,"G",@progbits,1175092228111723119,comdat
```
Comment (MaskRay) [inline lld/test/ELF/Inputs/dwarf5-debug-names-type-comdat.s:1]: For newer tests we tend to move away from `Inputs/`. dwarf5-debug-names-type-comdat.s and the DWARF64 counterpart can be placed into `x86-64-dwarf5-64-debug-names-type-comdat.test`. Search for `.ifdef ERR` and `llvm-mc --defsym`. We can use some assembly preprocessing to deduplicate the input.

File: lld/test/ELF/x86-64-dwarf5-debug-names-type-comdat.test [hunk 8]
```diff
@@ -0,0 +1,16 @@
+// REQUIRES: x86
```
Comment (MaskRay) [inline lld/test/ELF/x86-64-dwarf5-debug-names-type-comdat.test:1]: In newer tests we make the comment (non-RUN non-CHECK) marker outstanding. We typically use `# RUN:` `# CHECK:` and `## Regular comments.`

File: lld/test/ELF/x86-64-dwarf5-64-debug-names-type-comdat.test [hunk 9]
```diff
@@ -0,0 +1,18 @@
+// REQUIRES: x86
```
Comment (MaskRay) [inline lld/test/ELF/x86-64-dwarf5-64-debug-names-type-comdat.test:1]: We should place the DWARF64 test into the other file.

https://maskray.me/blog/2021-08-08-toolchain-testing#the-test-checks-at-the-wrong-layer#i-dont-know-an-existing-test-can-be-enhanced applies

**Review by MaskRay — COMMENTED** [review]
Code-wise this is close, but I want to see #68131 landed so that I can play with .debug_names tombstone values.

My previous comment still applies: the description should describe the choice of UINT32_MAX and UINT64_MAX, so that a curious reader doesn't have to read several external links to understand the choice.

**Comment by dwblaikie** [comment]
> My previous comment still applies: the description should describe the choice of UINT32_MAX and UINT64_MAX, so that a curious reader doesn't have to read several external links to understand the choice.

Is there a way we could avoid hardcoding this, and just do "max of whatever the relocation size is"? (some earlier versions of this patch had switches over relocation types to try to do this - which seemed a bit verbose/brittle/repetitive - but if there was some way to generalize it tidily, I think that'd be ideal)

**Comment by ayermolo** [comment]
> > My previous comment still applies: the description should describe the choice of UINT32_MAX and UINT64_MAX, so that a curious reader doesn't have to read several external links to understand the choice.
> 
> Is there a way we could avoid hardcoding this, and just do "max of whatever the relocation size is"? (some earlier versions of this patch had switches over relocation types to try to do this - which seemed a bit verbose/brittle/repetitive - but if there was some way to generalize it tidily, I think that'd be ideal)

That is already what is happening. We always extend to 64bit MAX, and value is set by relocation type. 32 bit for DWARF32, and 64bit for DWARF64.

**Comment by ayermolo** [comment]
@MaskRay OK, cleaned up tests.

**Comment by ayermolo** [comment]
@MaskRay ping :)

**Comment by MaskRay** [comment]
Sorry for not chiming in earlier. I think the current description is
probably inadequate, so a future reader would have a hard time to
analyze this use case.

With `git log --grep debug_names`, I see that #73872 has recently landed
(Dec 4) and now the curious (like I) can redo the experiment.
I am still expecting a concrete example in the patch description (first comment of this PR). https://github.com/llvm/llvm-project/pull/70701#issuecomment-1793910821

I think the description can be rephased like the following:

```
#73872 added support for ...

% cat a.cc
...
% cat b.cc
...
% clang++ -g -gpubnames -fdebug-types-section -fuse-ld=lld a.cc b.cc -o old
% llvm-dwarfdump -v old
....
.... // add a comment that an entry should use the tombstone value
instead.

https://dwarfstd.org/issues/231013.1.html proposes that ...
This patch implements the idea. The new llvm-dwarfdump output will look like:

% llvm-dwarfdump -v new
```

File: lld/ELF/InputSection.cpp [hunk 10]
```diff
@@ -886,16 +885,25 @@ void InputSection::relocateNonAlloc(uint8_t *buf, ArrayRef<RelTy> rels) {
   const unsigned bits = sizeof(typename ELFT::uint) * 8;
   const TargetInfo &target = *elf::target;
   const bool isDebug = isDebugSection(*this);
-  const bool isDebugLocOrRanges =
-      isDebug && (name == ".debug_loc" || name == ".debug_ranges");
   const bool isDebugLine = isDebug && name == ".debug_line";
-  std::optional<uint64_t> tombstone;
+  std::optional<uint64_t> tombstone = std::nullopt;
```
Comment (MaskRay) [inline lld/ELF/InputSection.cpp:889]: An optional defaults to std::nullopt. Just default initialize it.

File: lld/test/ELF/x86-64-dwarf5-64-debug-names-type-comdat.test [hunk 11]
```diff
@@ -0,0 +1,97 @@
+# REQUIRES: x86
+
+# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux --defsym DWARF32=1 %s -o %t.o
+# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux --defsym DWARF32=1 %s -o %t1.o
+# RUN: ld.lld %t.o %t1.o -o %t1
+# RUN: llvm-objdump -s %t1 | FileCheck %s --check-prefix=CHECK32
+
+# Test checks that LLD tombstones TU section that was de-duplicated using COMDAT to the maxium value.
```
Comment (MaskRay) [inline lld/test/ELF/x86-64-dwarf5-64-debug-names-type-comdat.test:8]: For newer tests we use `## ` for non-RUN non-CHECK comments.

File: lld/test/ELF/x86-64-dwarf5-64-debug-names-type-comdat.test [hunk 12]
```diff
@@ -0,0 +1,97 @@
+# REQUIRES: x86
+
+# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux --defsym DWARF32=1 %s -o %t.o
+# RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux --defsym DWARF32=1 %s -o %t1.o
```
Comment (MaskRay) [inline lld/test/ELF/x86-64-dwarf5-64-debug-names-type-comdat.test:4]: You can do `ld.lld %t.o %t.o`.

`-triple=x86_64` since there is no Linux specific stuff. We omit the OS to essentially say the test is generic to all ELF OSes.

File: lld/ELF/InputSection.cpp [hunk 13]
```diff
@@ -20,8 +20,7 @@
 #include "llvm/Support/Compression.h"
 #include "llvm/Support/Endian.h"
 #include "llvm/Support/xxhash.h"
-#include <algorithm>
-#include <mutex>
```
Comment (MaskRay) [inline lld/ELF/InputSection.cpp:24]: We should keep the headers. mutex is used by std::mutex in this file.

**Comment by MaskRay** [comment]
I think I'd implement the logic quite differently. Let me put up a patch...

**Comment by MaskRay** [comment]
> #73872 added support for ...
> .debug_names references local TUs in COMDAT .debug_info.
> When the referenced .debug_info is non-prevailing, we leave a value of zero, which is a valid offset within the .debug_info section.
> 
> % cat a.cc
> ...
> % cat b.cc
> ...
> % clang++ -g -gpubnames -fdebug-types-section -fuse-ld=lld a.cc b.cc -o old
> % llvm-dwarfdump -v old
> ...
> Local Type Unit Offsets [
>   LocalTU[0]: 0x00000000  // this is a valid offset within the .debug_info section
> ...
> 
> https://dwarfstd.org/issues/231013.1.html proposes that we use a tombstone value instead.
> This patch implements the idea. The new llvm-dwarfdump output will look like:
> 
> % llvm-dwarfdump -v new
> ...
> Local Type Unit Offsets [
>   LocalTU[0]: 0xffffffff  // tombstone value
> ...

I have created #74686 as an alternative. Thank you for your and @dwblaikie's descriptions to improve my .debug_names understanding :)
